language/llama3-405b/SUT_VLLM.py

```diff
@@ -0,0 +1,293 @@
+            "min_tokens": 20
+        }
+        self.sampling_params = SamplingParams(**gen_kwargs)
+        # self.sampling_params.all_stop_token_ids.add(self.model.get_tokenizer().eos_token_id)
+
+        self.num_workers = workers
+        self.worker_threads = [None] * self.num_workers
+        self.query_queue = queue.Queue()
+
+        self.use_cached_outputs = use_cached_outputs
+        self.sample_counter = 0
+        self.sample_counter_lock = threading.Lock()
+
+    def start(self):
+        # Create worker threads
+        for j in range(self.num_workers):
+            worker = threading.Thread(target=self.process_queries)
+            worker.start()
+            self.worker_threads[j] = worker
+
+    def stop(self):
+        for _ in range(self.num_workers):
+            self.query_queue.put(None)
+
+        for worker in self.worker_threads:
+            worker.join()
+
+    def process_queries(self):
+        """Processor of the queued queries. User may choose to add batching logic"""
+
+        while True:
+            qitem = self.query_queue.get()
+            if qitem is None:
+                break
+
+            query_ids = [q.index for q in qitem]
+
+            tik1 = time.time()
+
+            input_ids_tensor = [
+                self.data_object.input_ids[q.index][:256] for q in qitem]
+
+            tik2 = time.time()
+            outputs = self.model.generate(
+                prompt_token_ids=input_ids_tensor, sampling_params=self.sampling_params
+            )
+            pred_output_tokens = []
+            for output in outputs:
+                pred_output_tokens.append(list(output.outputs[0].token_ids))
+            tik3 = time.time()
+
+            processed_output = self.data_object.postProcess(
+                pred_output_tokens,
+                query_id_list=query_ids,
+            )
+
+        for i in range(len(qitem)):
+            n_tokens = processed_output[i].shape[0]
+            response_array = array.array(
+                "B", processed_output[i].tobytes())
+            bi = response_array.buffer_info()
+            response = [
+                lg.QuerySampleResponse(
+                    qitem[i].id,
+                    bi[0],
+                    bi[1],
+                    n_tokens)]
+            lg.QuerySamplesComplete(response)
+
+        tok = time.time()
+
+        with self.sample_counter_lock:
+            self.sample_counter += len(qitem)
+            print(f"Samples run: {self.sample_counter}")
+            if tik1:
+                print(f"\tBatchMaker time: {tik2 - tik1}")
+                print(f"\tInference time: {tik3 - tik2}")
+                print(f"\tPostprocess time: {tok - tik3}")
+                print(f"\t==== Total time: {tok - tik1}")
+
+    def load_model(self):
+        print("Loading model...")
+        self.model = LLM(
+            self.model_path,
+            dtype=self.dtype,
+            tensor_parallel_size=self.tensor_parallel_size,
```

Comment:
Is there a reason we are not using AsyncLLMEngine here? It may be more efficient since it will support continous batching